perspective_reddit_bot/check_mod_actions.py

```diff
@@ -51,66 +51,163 @@
 FILENAME_OUTPUT_PREFIX = 'modactions'
 
 
-def write_moderator_actions(reddit,
-                            record,
-                            id_key,
-                            timestamp_key,
-                            output_path,
-                            hours_to_wait,
-                            has_mod_creds):
-  create_time_utc = datetime.strptime(record[timestamp_key], '%Y%m%d_%H%M%S')
-  check_time_utc = create_time_utc + timedelta(hours=hours_to_wait)
-  wait_until(check_time_utc)
-
-  record['action_checked_utc'] = log_subreddit_comments.now_timestamp()
-  status_fields = fetch_reddit_comment_status(reddit, record[id_key],
-                                              has_mod_creds)
-  if status_fields is not None:
-    record.update(status_fields)
-
-  log_subreddit_comments.append_records(output_path, [record])
-
-
-def fetch_reddit_comment_status(reddit, comment_id, has_mod_creds):
+# TODO: This implementation of batching is a bit complex: read_lines yields
+# sentinel values at regular intervals so that read_batches is able to yield
+# batches within max_batch_delay. The issue is that when reading from the end of
+# the file, there could be an indeterminate amount of time until the next line,
+# which would cause accumulated lines to be "stranded".
+#
+# This can almost certainly be done in a simpler way...
+
+def read_lines(input_path, stop_at_eof, yield_every_period, sentinel):
+  """Read lines from file.
+
+  Args:
+    input_path: (str) where to read from
+    stop_at_eof: (bool) returns reaching end of file if true.
+    yield_every_period: (timedelta) yields sentinel value every time period if
+        stream doesn't have contents.
+    sentinel: value yielded every yield_every_period
+
+  Yields: lines from input_path, or sentinel values
+  """
+  with open(input_path) as f:
+    last_time = datetime.utcnow()
+    while True:
+      now = datetime.utcnow()
+      where = f.tell()
+      line = f.readline()
+      if line:
+        print('.', end='')
+        sys.stdout.flush()
+        yield line
+        last_time = now
+      elif stop_at_eof:
+        print('\nread_lines stopping due to stop_at_eof.')
+        return
+      elif now - last_time > yield_every_period:
+        yield sentinel
+        last_time = now
+      else:
+        print('\n(eof, waiting {})'.format(yield_every_period))
+        time.sleep(yield_every_period.total_seconds())
+        f.seek(where)
+
+
+_PRAW_BATCH_SIZE = 100
+
+
+def read_batches(input_path, stop_at_eof, max_batch_delay):
```

Comment:
Does this construction actually guarantee that you yield the batch under the max_batch_delay? Couldn't one comment sneak in just before the delay and then you wait another max_batch_delay seconds before yielding the sentinel?

I suppose you could make yield_every_period = max_batch_delay / 2 to fix this? Let me know if I'm thinking about this wrong.

Reply:
I think you're correct. I had `max_batch_delay /  2` at some point, but later removed it because I ...  didn't want to add a  comment why it was needed :-D plus with max_batch_delay << hours_to_wait, it's not a big deal.

But I'll add the /  2!